### PR TITLE
Implement Index Lifecycle APIs

### PR DESCRIPTION
**Pull request by codebrain**
Would like to discuss the Put Lifecycle API methods on the client.

### CODE REVIEW

File: src/Nest/CommonAbstractions/Infer/Policy/Policy.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+﻿using System;
+using System.Globalization;
+using Elasticsearch.Net;
+
+namespace Nest
+{
+	public class PolicyId : IUrlParameter, IEquatable<PolicyId>
```
Comment (russcam) [inline src/Nest/CommonAbstractions/Infer/Policy/Policy.cs]: Can we use `Id` for this?

File: src/Nest/XPack/Ilm/Actions/AllocateAction.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
+using System;
+using System.Collections.Generic;
+using Newtonsoft.Json;
+
+namespace Nest
+{
+	public interface IAllocateLifecycleAction : ILifecycleAction
```
Comment (russcam) [inline src/Nest/XPack/Ilm/Actions/AllocateAction.cs]: Are there XML docs that can be added?

File: src/Nest/XPack/Ilm/Actions/ForceMergeAction.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+using Newtonsoft.Json;
+
+namespace Nest
+{
+	public interface IForceMergeLifecycleAction : ILifecycleAction
+	{
+		[JsonProperty("max_num_segments")]
+		int? MaximumNumberSegments { get; set; }
```
Comment (russcam) [inline src/Nest/XPack/Ilm/Actions/ForceMergeAction.cs]:
```suggestion
		int? MaximumNumberOfSegments { get; set; }
```

File: src/Nest/XPack/Ilm/Actions/ForceMergeAction.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+	}
+
+	public class ForceMergeLifecycleAction : IForceMergeLifecycleAction
+	{
+		public int? MaximumNumberSegments { get; set; }
```
Comment (russcam) [inline src/Nest/XPack/Ilm/Actions/ForceMergeAction.cs]:
```suggestion
		public int? MaximumNumberOfSegments { get; set; }
```

File: src/Nest/XPack/Ilm/Actions/ForceMergeAction.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	}
+
+	public class ForceMergeLifecycleActionDescriptor : DescriptorBase<ForceMergeLifecycleActionDescriptor, IForceMergeLifecycleAction>, IForceMergeLifecycleAction
+	{
+		int? IForceMergeLifecycleAction.MaximumNumberSegments { get; set; }
```
Comment (russcam) [inline src/Nest/XPack/Ilm/Actions/ForceMergeAction.cs]:
```suggestion
		int? IForceMergeLifecycleAction.MaximumNumberOfSegments { get; set; }
```

File: src/Nest/XPack/Ilm/Actions/ForceMergeAction.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+		public ForceMergeLifecycleActionDescriptor MaximumNumberSegments(int? maximumNumberSegments)
```
Comment (russcam) [inline src/Nest/XPack/Ilm/Actions/ForceMergeAction.cs]:
```suggestion
		public ForceMergeLifecycleActionDescriptor MaximumNumberOfSegments(int? maximumNumberOfSegments)
```

File: src/Nest/XPack/Ilm/Actions/ForceMergeAction.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+			=> Assign(maximumNumberSegments, (a, v) => a.MaximumNumberSegments = maximumNumberSegments);
```
Comment (russcam) [inline src/Nest/XPack/Ilm/Actions/ForceMergeAction.cs]:
```suggestion
			=> Assign(maximumNumberSegments, (a, v) => a.MaximumNumberOfSegments = maximumNumberOfSegments);
```

File: src/Nest/XPack/Ilm/Actions/RolloverAction.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
+using Newtonsoft.Json;
+
+namespace Nest
+{
+	public interface IRolloverLifecycleAction : ILifecycleAction
```
Comment (russcam) [inline src/Nest/XPack/Ilm/Actions/RolloverAction.cs]: Are there XML docs that can be added?

File: src/Nest/XPack/Ilm/Actions/SetPriorityAction.cs [hunk 9]
```diff
@@ -0,0 +1,22 @@
+using Newtonsoft.Json;
+
+namespace Nest
+{
+	public interface ISetPriorityLifecycleAction : ILifecycleAction
```
Comment (russcam) [inline src/Nest/XPack/Ilm/Actions/SetPriorityAction.cs]: Are there XML docs that can be added?

File: src/Nest/XPack/Ilm/Actions/ShrinkAction.cs [hunk 10]
```diff
@@ -0,0 +1,22 @@
+using Newtonsoft.Json;
+
+namespace Nest
+{
+	public interface IShrinkLifecycleAction : ILifecycleAction
```
Comment (russcam) [inline src/Nest/XPack/Ilm/Actions/ShrinkAction.cs]: Are there XML docs that can be added?

File: src/Nest/XPack/Ilm/Actions/UnfollowAction.cs [hunk 11]
```diff
@@ -0,0 +1,12 @@
+namespace Nest
+{
+	public interface IUnfollowLifecycleAction : ILifecycleAction { }
```
Comment (russcam) [inline src/Nest/XPack/Ilm/Actions/UnfollowAction.cs]: Are there XML docs that can be added?

File: src/Nest/XPack/Ilm/DeleteLifecycle/ElasticClient-IlmDeleteLifecycle.cs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+namespace Nest
+{
+	public partial interface IElasticClient
+	{
+		/// <summary>
+		/// Deletes a lifecycle policy.
+		/// Deletes the specified lifecycle policy definition. You cannot delete policies that are currently in use.
+		/// If the policy is being used to manage any indices, the request fails and returns an error.
+		/// </summary>
+		IIlmDeleteLifecycleResponse IlmDeleteLifecycle(PolicyId policyId, Func<IlmDeleteLifecycleDescriptor, IIlmDeleteLifecycleRequest> selector = null);
```
Comment (russcam) [inline src/Nest/XPack/Ilm/DeleteLifecycle/ElasticClient-IlmDeleteLifecycle.cs]: Not sure about the `Ilm` prefix in the high level client; the prefix has typically not been used, except in cases where it was required to disambiguate methods. Are there any methods that would conflict if the prefix were removed?

File: src/Nest/XPack/Ilm/ExplainLifecycle/IlmExplainLifecycleResponse.cs [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+		public IReadOnlyDictionary<string, LifecycleExplain> Indices { get; internal set; }
+	}
+
+	public class LifecycleExplain
+	{
+		[JsonProperty("index")]
+		public IndexName Index { get; internal set; }
+
+		/// <summary>
+		/// Shows if the index is being managed by ILM. If the index is not managed by ILM the other fields will not be shown.
```
Comment (russcam) [inline src/Nest/XPack/Ilm/ExplainLifecycle/IlmExplainLifecycleResponse.cs]:
```suggestion
		/// Shows if the index is being managed by Index Lifecycle Management (ILM). If the index is not managed by ILM the other fields will not be shown.
```

File: src/Nest/XPack/Ilm/GetStatus/OperationMode.cs [hunk 14]
```diff
@@ -0,0 +1,14 @@
+using System.Runtime.Serialization;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+
+namespace Nest
+{
+	[JsonConverter(typeof(StringEnumConverter))]
+	public enum OperationMode
```
Comment (russcam) [inline src/Nest/XPack/Ilm/GetStatus/OperationMode.cs]: Maybe `LifecycleOperationMode`?

File: src/Nest/XPack/Ilm/MoveToStep/IlmMoveToStepRequest.cs [hunk 15]
```diff
@@ -0,0 +1,56 @@
+using System;
+using Newtonsoft.Json;
+
+namespace Nest
+{
+	public partial interface IIlmMoveToStepRequest
```
Comment (russcam) [inline src/Nest/XPack/Ilm/MoveToStep/IlmMoveToStepRequest.cs]: Are there XML docs that can be added?

File: src/Nest/CommonAbstractions/Infer/Policy/Policy.cs (hunk 1 above)
Comment (codebrain) [reply src/Nest/CommonAbstractions/Infer/Policy/Policy.cs]: Sadly, Id allows `long` and this is constrained to `string`. We have a ticket open for consolidating these custom types, maybe look at this then? I imagine we might well end up with `Id` and `IdString` or something similar.

File: src/Nest/XPack/Ilm/GetStatus/OperationMode.cs (hunk 14 above)
Comment (codebrain) [reply src/Nest/XPack/Ilm/GetStatus/OperationMode.cs]: ++ done.

File: src/Nest/XPack/Ilm/Actions/AllocateAction.cs (hunk 2 above)
Comment (codebrain) [reply src/Nest/XPack/Ilm/Actions/AllocateAction.cs]: Done.

File: src/Nest/XPack/Ilm/Actions/RolloverAction.cs (hunk 8 above)
Comment (codebrain) [reply src/Nest/XPack/Ilm/Actions/RolloverAction.cs]: Done.

File: src/Nest/XPack/Ilm/Actions/ShrinkAction.cs (hunk 10 above)
Comment (codebrain) [reply src/Nest/XPack/Ilm/Actions/ShrinkAction.cs]: Done.

File: src/Nest/XPack/Ilm/Actions/SetPriorityAction.cs (hunk 9 above)
Comment (codebrain) [reply src/Nest/XPack/Ilm/Actions/SetPriorityAction.cs]: Done.

File: src/Nest/XPack/Ilm/Actions/UnfollowAction.cs (hunk 11 above)
Comment (codebrain) [reply src/Nest/XPack/Ilm/Actions/UnfollowAction.cs]: Done.

File: src/Nest/XPack/Ilm/MoveToStep/IlmMoveToStepRequest.cs (hunk 15 above)
Comment (codebrain) [reply src/Nest/XPack/Ilm/MoveToStep/IlmMoveToStepRequest.cs]: Done.

File: src/Nest/CommonAbstractions/Infer/Policy/Policy.cs (hunk 1 above)
Comment (codebrain) [reply src/Nest/CommonAbstractions/Infer/Policy/Policy.cs]: Will address in https://github.com/elastic/elasticsearch-net/issues/3672

File: src/Nest/XPack/Ilm/DeleteLifecycle/ElasticClient-IlmDeleteLifecycle.cs (hunk 12 above)
Comment (codebrain) [reply src/Nest/XPack/Ilm/DeleteLifecycle/ElasticClient-IlmDeleteLifecycle.cs]: I've renamed everything to remove the Ilm prefix. I believe the new naming is better and should address your feedback sufficiently.

**Pull request merged**